### ISSUE TITLE
Allow member type of map/array to be union of basic types

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/nballerina#328]
This means we allow type descriptors like

```
int[]
map<int>
map<int|string>
(int|string?)[]
```

Implementationally:

- we maintain a single representation of maps/arrays where values are tagged
- map/array values have an inherent type that can be represented as a bit vector of basic types
- we also need to deal with different filler values

This lays the ground for the error type #167 (since that requires implementing `any[]` as having a member type of any basic type other than error).

This depends on #327.

Implementation follow-on from this is #329.

**Comment by jclark** [comment]
One parsing complication is that when we see a `(` at the beginning of a statement we don't know if we have a method call statement of a local variable declaration.

**Comment by jclark** [comment]
Also need to deal with `array:push` being generic.

**Comment by jclark** [comment]
Also need to deal with relational operators being allowed when there is an ordered type (e.g. `int[]`) to which both operands belong.

**Comment by jclark** [comment]
Type narrowing will need to use read-only difference operation.

**Comment by jclark** [comment]
We have to support `==` and `!=` when the member types are simple or strings, but we don't yet need to support comparison of values with cycles (because we don't yet allow e.g. `map<map<int>>`).

Note that this also requires distinguishing `==` and `===`, which we have to do for `float` anyway.

**Comment by jclark** [comment]
This also requires implementing compound assignment when LHS is `x[i]`.

**Comment by jclark** [comment]
Still to do:

- [x] make member access expressions work: make sure untagging happens; fix verify
- [x] make sure list/mapping construction is working (need to fix verify)
- [x] Implement == and != when member type is simple or string
- [x] make ListSet/MappingSet work in front-end/verify
- [x] add inherent type to map and use in MappingConstruct/MappingSet insns
- [x] add inherent type to list and use in ListConstruct/ListSet insns
- [x] use inherent type to choose filler value for list
- [x] type cast/test should check inherent type (but only when necessary)

Separate issues that are required for this:

- fix array:push #409
- extend CompareInsn to arrays #410
- support parentheses in array type syntax `(any|error)[]` #411
- support `v[k] op= expr;` #415
- handle narrowing #420

**Comment by jclark** [comment]
Everything remaining is a separate issue, so closing this.
